### PR TITLE
Zaimanhua: fix image url expire problem & incorrect chapter upload time

### LINKED ISSUE #5414: zaimanhua/再漫画 BUG
### Source information

zaimanhua 1.4.1

### Source language

Chinese

### Steps to reproduce

正常阅读漫画

### Expected behavior

漫画图片正常加载

### Actual behavior

每部阅读几十页后就有高概率出现一页加载失败，并且后续所有页数都是加载失败，且点击重试也是加载失败，只有重启mihon app后才可以重新正常加载漫画图片。其他插件都不会这样，只有再漫画插件频繁出现这个问题，缓存也都清过了。
Screenshot_2024-10-07-03-24-57-079_app mihon

### Mihon/Tachiyomi version

mihon 0.16.5

### Android version

Android 13

### Other details

_No response_

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a short but informative title.
- [X] I have updated the app to version **[0.15.3](https://github.com/mihonapp/mihon/releases/latest)**.
- [X] I have updated all installed extensions.
- [X] I have tried the [troubleshooting guide](https://tachiyomi.org/help/guides/troubleshooting/).
- [X] If this is an issue with the app itself, I should be opening an issue in the [app repository](https://github.com/tachiyomiorg/tachiyomi/issues/new/choose).
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by zhongfly**
Close #5414 

Checklist:

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [ ] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [ ] Have removed `web_hi_res_512.png` when adding a new extension

### CODE REVIEW

**Comment by zhongfly** [comment]
In some specific cases, the image was not downloaded immediately after calling fetchImageUrl (e.g., the user interrupted reading or cancels the download). If the same chapter open/download again, the imageUrls of these images is expired and will always fail to load.
And it is difficult to back out the request parameters used in fetchImageUrl through the expired imageUrl, which also makes the idea of intercepting the image load failure and then re-fetching the image link difficult to realize.

I've decided to ignore this edge situation for now.

File: src/zh/zaimanhua/src/eu/kanade/tachiyomi/extension/zh/zaimanhua/Zaimanhua.kt [hunk 1]
```diff
@@ -129,24 +132,34 @@ class Zaimanhua : HttpSource(), ConfigurableSource {
...
+
+    override fun fetchImageUrl(page: Page): Observable<String> {
+        val response = client.newCall(GET(page.url, apiHeaders, USE_CACHE)).execute()
+        val result = response.parseAs<ResponseDto<DataWrapperDto<ChapterImagesDto>>>()
+        if (result.errmsg.isNotBlank()) {
+            throw Exception(result.errmsg)
+        } else {
+            return Observable.just(result.data.data!!.images[page.index])
+        }
+    }
```
Comment (AwkwardPeak7) [inline src/zh/zaimanhua/src/eu/kanade/tachiyomi/extension/zh/zaimanhua/Zaimanhua.kt]: consider using combination of `imageUrlRequest` and `ImageUrlParse` instead

**Review by AwkwardPeak7 — CHANGES_REQUESTED** [review]

**Comment by AwkwardPeak7** [comment]
>  And it is difficult to back out the request parameters used in fetchImageUrl through the expired imageUrl, which also makes the idea of intercepting the image load failure and then re-fetching the image link difficult to realize.

have you tried saving the request parms etc in the image url fragment (maybe as a json string) and using okhttp interceptors instead?

File: src/zh/zaimanhua/src/eu/kanade/tachiyomi/extension/zh/zaimanhua/Zaimanhua.kt (hunk 1 above)
Comment (zhongfly) [reply src/zh/zaimanhua/src/eu/kanade/tachiyomi/extension/zh/zaimanhua/Zaimanhua.kt]: In `ImageUrlParse`, `page.index` is invalid.

**Comment by zhongfly** [comment]
> have you tried saving the request parms etc in the image url fragment (maybe as a json string) and using okhttp interceptors instead?

Done

File: src/zh/zaimanhua/src/eu/kanade/tachiyomi/extension/zh/zaimanhua/Zaimanhua.kt [hunk 2]
```diff
@@ -129,24 +141,48 @@ class Zaimanhua : HttpSource(), ConfigurableSource {
...
+        val response = chain.proceed(request)
+        val fragment = request.url.fragment
+        if (response.isSuccessful || request.url.host != "images.zaimanhua.com" || fragment == null) return response
+        response.close()
+
+        val params = json.decodeFromString<ImageRetryParamsDto>(fragment)
+        val pageListResponse = client.newCall(pageListApiRequest(params.url)).execute()
+        val result = pageListResponse.parseAs<ResponseDto<DataWrapperDto<ChapterImagesDto>>>()
+        if (result.errmsg.isNotBlank()) {
+            throw Exception(result.errmsg)
```
Comment (AwkwardPeak7) [inline src/zh/zaimanhua/src/eu/kanade/tachiyomi/extension/zh/zaimanhua/Zaimanhua.kt]: only throw `IOException` from interceptors

Comment (zhongfly) [reply src/zh/zaimanhua/src/eu/kanade/tachiyomi/extension/zh/zaimanhua/Zaimanhua.kt]: fixed

**Review by AwkwardPeak7 — APPROVED** [review]

**Pull request merged**